### PR TITLE
HDDS-1318. Fix MalformedTracerStateStringException on DN logs. Contributed by Xiaoyu Yao.

### PR DESCRIPTION
**Pull request by xiaoyuyao**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 39 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 70 | Maven dependency ordering for branch |
| +1 | mvninstall | 1250 | trunk passed |
| +1 | compile | 1277 | trunk passed |
| +1 | checkstyle | 210 | trunk passed |
| +1 | mvnsite | 174 | trunk passed |
| +1 | shadedclient | 1178 | branch has no errors when building and testing our client artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: hadoop-ozone/integration-test |
| +1 | findbugs | 122 | trunk passed |
| +1 | javadoc | 103 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 20 | Maven dependency ordering for patch |
| +1 | mvninstall | 94 | the patch passed |
| +1 | compile | 1018 | the patch passed |
| +1 | javac | 1018 | the patch passed |
| +1 | checkstyle | 200 | the patch passed |
| +1 | mvnsite | 119 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 702 | patch has no errors when building and testing our client artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: hadoop-ozone/integration-test |
| +1 | findbugs | 130 | the patch passed |
| +1 | javadoc | 102 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 30 | client in the patch passed. |
| +1 | unit | 95 | common in the patch passed. |
| -1 | unit | 859 | integration-test in the patch failed. |
| +1 | asflicense | 45 | The patch does not generate ASF License warnings. |
| | | 7730 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.ozoneimpl.TestSecureOzoneContainer |
|   | hadoop.ozone.container.ozoneimpl.TestOzoneContainer |
|   | hadoop.ozone.container.TestContainerReplication |
|   | hadoop.ozone.container.metrics.TestContainerMetrics |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-641/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/641 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux a8ba333b4d6a 3.13.0-153-generic #203-Ubuntu SMP Thu Jun 14 08:52:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3f6d6d2 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/1/artifact/out/patch-unit-hadoop-ozone_integration-test.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/1/testReport/ |
| Max. process+thread count | 4096 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/client hadoop-hdds/common hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/tracing/StringCodec.java [hunk 1]
```diff
@@ -25,12 +25,15 @@
 import io.jaegertracing.internal.exceptions.TraceIdOutOfBoundException;
 import io.jaegertracing.spi.Codec;
 import io.opentracing.propagation.Format;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * A jaeger codec to save the current tracing context t a string.
```
Comment (ajayydv) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/tracing/StringCodec.java]: can we fix this typo as well?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/ozShell/TestOzoneShell.java [hunk 2]
```diff
@@ -919,13 +926,19 @@ public void testGetKey() throws Exception {
...
+        logs.getOutput().contains("MalformedTracerStateString"));
+    logs.clearOutput();
 
     String tmpPath = baseDir.getAbsolutePath() + "/testfile-"
         + UUID.randomUUID().toString();
     String[] args = new String[] {"key", "get",
         url + "/" + volumeName + "/" + bucketName + "/" + keyName,
         tmpPath};
     execute(shell, args);
+    assertFalse("get key without malformed tracing",
```
Comment (ajayydv) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/ozShell/TestOzoneShell.java]: Shall we check the case when it is malformed?

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java [hunk 3]
```diff
@@ -217,7 +220,21 @@ public XceiverClientReply sendCommand(
       ContainerCommandRequestProto request, List<DatanodeDetails> excludeDns)
       throws IOException {
     Preconditions.checkState(HddsUtils.isReadOnly(request));
-    return sendCommandWithRetry(request, excludeDns);
+    return sendCommandWithTraceIDAndRetry(request, excludeDns);
```
Comment (ajayydv) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java]: shall we do this only when tracing is enabled?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/ozShell/TestOzoneShell.java (hunk 2 above)
Comment (xiaoyuyao) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/ozShell/TestOzoneShell.java]: malformed trace can be easily reproed without the production code fix when getKey is called (e.g., in this test).

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java (hunk 3 above)
Comment (xiaoyuyao) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java]: Unfortunately, I'm not aware of a switch to turn tracing off globally. That will be a much bigger change than the scope of this ticket.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/tracing/StringCodec.java (hunk 1 above)
Comment (xiaoyuyao) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/tracing/StringCodec.java]: sure, will fix it in next commit.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/ozShell/TestOzoneShell.java (hunk 2 above)
Comment (xiaoyuyao) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/ozShell/TestOzoneShell.java]: additional unit test added.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 28 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 61 | Maven dependency ordering for branch |
| +1 | mvninstall | 997 | trunk passed |
| +1 | compile | 921 | trunk passed |
| +1 | checkstyle | 178 | trunk passed |
| +1 | mvnsite | 137 | trunk passed |
| +1 | shadedclient | 997 | branch has no errors when building and testing our client artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: hadoop-ozone/integration-test |
| +1 | findbugs | 110 | trunk passed |
| +1 | javadoc | 85 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for patch |
| +1 | mvninstall | 95 | the patch passed |
| +1 | compile | 946 | the patch passed |
| +1 | javac | 946 | the patch passed |
| +1 | checkstyle | 185 | the patch passed |
| +1 | mvnsite | 116 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 616 | patch has no errors when building and testing our client artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: hadoop-ozone/integration-test |
| +1 | findbugs | 119 | the patch passed |
| +1 | javadoc | 104 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 35 | client in the patch passed. |
| +1 | unit | 77 | common in the patch passed. |
| -1 | unit | 1456 | integration-test in the patch failed. |
| +1 | asflicense | 38 | The patch does not generate ASF License warnings. |
| | | 7273 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.om.TestOzoneManagerHA |
|   | hadoop.ozone.container.TestContainerReplication |
|   | hadoop.ozone.container.metrics.TestContainerMetrics |
|   | hadoop.ozone.container.ozoneimpl.TestSecureOzoneContainer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-641/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/641 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 089b0d54691b 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a4cd75e |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/2/artifact/out/patch-unit-hadoop-ozone_integration-test.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/2/testReport/ |
| Max. process+thread count | 3674 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/client hadoop-hdds/common hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by xiaoyuyao** [comment]
Some of the test failures are related to this change. I'm looking into it.

**Comment by xiaoyuyao** [comment]
The current pattern of export and import traceID over the wire (GRPC/Ratis) is via the TraceID in the protocol message. But this changed the original traceID which the test expect to be the same, which is not true with the new yaeger based tracing where the parent:children:... are appended along the invocation chain. Will remove the invalid verification from those tests, cc: @elek .

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 28 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 10 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 28 | Maven dependency ordering for branch |
| +1 | mvninstall | 994 | trunk passed |
| +1 | compile | 957 | trunk passed |
| +1 | checkstyle | 193 | trunk passed |
| +1 | mvnsite | 170 | trunk passed |
| +1 | shadedclient | 1092 | branch has no errors when building and testing our client artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: hadoop-ozone/integration-test |
| +1 | findbugs | 125 | trunk passed |
| +1 | javadoc | 117 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for patch |
| +1 | mvninstall | 100 | the patch passed |
| +1 | compile | 904 | the patch passed |
| +1 | javac | 904 | the patch passed |
| +1 | checkstyle | 231 | the patch passed |
| +1 | mvnsite | 150 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 665 | patch has no errors when building and testing our client artifacts. |
| 0 | findbugs | 0 | Skipped patched modules with no Java source: hadoop-ozone/integration-test |
| +1 | findbugs | 137 | the patch passed |
| +1 | javadoc | 116 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 36 | client in the patch passed. |
| +1 | unit | 77 | common in the patch passed. |
| -1 | unit | 653 | integration-test in the patch failed. |
| +1 | asflicense | 52 | The patch does not generate ASF License warnings. |
| | | 6767 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.ozShell.TestOzoneShell |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-641/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/641 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 21d34faea4eb 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9cd6619 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/3/artifact/out/patch-unit-hadoop-ozone_integration-test.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/3/testReport/ |
| Max. process+thread count | 4532 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/client hadoop-hdds/common hadoop-ozone/integration-test U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-641/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by xiaoyuyao** [comment]
The single test failure does not repro locally. Seems unrelated to this patch.

**Comment by ajayydv** [comment]
+1

**Comment by xiaoyuyao** [comment]
Thanks for the review @ajayydv , the checkstyle issue is not introduced by this patch. I've opened a separate JIRA: https://issues.apache.org/jira/browse/HDDS-1350 so that we can get a clean cherry-pick for ozone-0.4 here.

**Pull request merged**